shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -843,6 +882,62 @@ - (void)viewDidDisappear:(BOOL)animated {
   [super viewDidDisappear:animated];
 }
 
+- (void)viewWillTransitionToSize:(CGSize)size
+       withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator {
+  [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
+
+  // We interpolate the viewport metrics (size and paddings) during rotation transition, to address
+  // a bug with distorted aspect ratio.
+  // See: https://github.com/flutter/flutter/issues/16322
+  //
+  // For every `kRotationViewportMetricsUpdateInterval`, we send the metrics which is interpolated
+  // between the old metrics before the rotation transition, to the new metrics after the rotation
+  // transition.
+  //
+  // Currently it is using linear interpolation. Using non-linear ease-in/out interpolation may
+  // achieve better results. It may also help to send only rotation info (such as rotation duration)
+  // and perform the interpolation on the framework side, to reduce engine/framework communication.
+  // However, since flutter's drawing happens on the ui thread, which is not iOS main thread,
+  // there is no guarantee that the viewport metrics change is immediately taken effect, resulting
+  // in some amount of unavoidable distortion.
+
+  NSTimeInterval transitionDuration = coordinator.transitionDuration;
+  // Do not interpolate if zero transition duration.
+  if (transitionDuration == 0) {
+    return;
+  }
+
+  _isDuringRotationTransition = YES;
+
+  CGSize oldSize = self.view.bounds.size;
+  UIEdgeInsets oldPadding = self.view.safeAreaInsets;
+
+  __block double rotationProgress = 0;
+  // Timer is retained by the run loop, and will be released after invalidated.
+  [NSTimer
+      scheduledTimerWithTimeInterval:kRotationViewportMetricsUpdateInterval
```

Comment:
Maybe replace  kRotationViewportMetricsUpdateInterval with `1/[DisplayLinkManager displayRefreshRate]` for pro-motion devices.

Reply:
Probably not necessary during rotation since the device is moving. I've even tried 30fps - not too bad either

Reply:
i can test with promotion device

Reply:
Feel free to pull the branch, or can wait for me to fix my laptop's engine build (hopefully this week).

Reply:
Actually, let me see if I can expense one.